binary-option/README.md

```diff
@@ -0,0 +1,82 @@
+# Binary Option
+
+This protocol is a primitive version of a binary options. Participants can enter a long or short position depending on their conviction. (These sides are set completely arbitrarily). The eventual goal is to have a higher level program manage the pool and handle settlements with an oracle based voting approach. Every bet in the pool involves 2 parties (1 long and 1 short) each depositing some collateral. Because bets are made on binary event, the sum of collateral will be equal to a mutiple of some power of 10 (`10 ** N` where `N` is configurable).
+
+The module contains the Rust implementation of protocol as well as a Python client and test suite.
+
+Suppose we had a binary option on the winner of the 2021 NBA Finals (Phoenix Suns vs. Milwaulkee Bucks). At the time of writing this (July 9th, 2021), the moneyline spread is -190 Suns +170 Bucks. This backs out an implied probability of approximately 36% that the Bucks win the championship. Suppose our binary option was on the Bucks winning this series, and that it is denominated by some wrapped stablecoin WUSD (dollar pegged) where every contract settled to 10000 WUSD (`N = 4` corresponding to 1 cent granualrity). You observe that someone is willing to go short Bucks for 10 contracts at 3000 WUSD (less than the estimated probability of 36%). You can take on the opposite trade by buying 10 long contracts on the Bucks for 3000.
+
+This invokes a `Trade` instruction with size 10, buy_price 3000, and sell_price 7000. Note that these prices must sum to 10000. As part of the protocol, you transfer 30000 WUSD into the binary option and the counterparty deposits 70000 (assuming that both parties start with 0 position). In return, 10 long tokens (minted by the contract) are added to your account, and 10 short tokens are minted to your counterparty's account.
+
+Now suppose the Bucks win Game 3, and the estimated probablity of the Bucks winning the series jumps to 40%. You offer to sell all of your contracts for 4000 WUSD, and you get a buyer. Because you already hold long tokens, the contract will burn those existing tokens, and you are transfered 40000 WUSD from the pool. If your counterparty is currently short 1 contract, they pull out 6000 WUSD from the pool (exiting out of their short position) and deposit 9 * 4000 = 36000 WUSD (buying into their long position). In total, the pool collateral changes by -40000 + 36000 - 6000 = -10000 WUSD or exactly 1 contract! After the dust settles, you walk away with no position and a net profit of 10000 WUSD ($100).
```

Comment:
buy_price = 4000, sell_price 6000

In this case "sell_price" refers to the amount you need to deposit into the pool to enter a new short position. However because you already have long positions, it's the equivalent of liquidating those long positions for 4000. This is the most confusing concept, but I did the math (made sure every possible case makes sense from like a real betting perspective). You only need to pay that 6000 if you start out with no long position.

Convention is funny and often times very confusing. I can only imagine why US sports books usually use American odds (like +200 referring to a payout of 200$ for every 100$ deposited). People are bad at math so it's easier to think about things in terms of the payouts, but in reality these are all just markets on probabilities and IMO it's easier to think about contracts this way. If every transaction increments or decrements the pool by some multiple of 100 or whatever the contract size is (i.e. 10 ^ decimals), then you never have to worry about the pool being under-collateralized.

Reply:
Gotcha, that makes a lot of sense.  For my own education, what would the inputs be if the person instead wanted to go long *again* and buy 10 more long tokens?

Reply:
So then they would just buy at whatever the quoted price is (you can think of this is as just the implied probability).

Suppose Bucks traded at 6000, and person A reenters their long position starting from 0. They will deposit 60k in to the pool and the shorts will either buy in for 40k in the case that they have no position or they are already short. If they are long `n` tokens, they will receive a payment of 6000n (essentially this is like the seller receiving direct payment from person A for these tokens). If n < 10, then they will deposit (10 - n) * 4000.

Reply:
and if A just holds till settle (Bucks won!), A would receive a total of 100k and anyone who was short will have lost everything.

Reply:
Ok I think I'm finally understanding.  If I want to go long 10, I call `Trade` with some new long / short accounts, let's call those the A accounts.  Let's say I also go short 10, so I call `Trade` with a *new* set of long / short accounts, let's call those the B accounts.  What happens if I call `Trade` with my 10 longs from A and my 10 shorts from B against someone else?